whylog/tests/tests_config/test_get_search_range.py

```diff
@@ -46,33 +96,260 @@ def setUpClass(cls):
+                    'right_bound': self.effect_time
+                }
+            },
+        }
+
+        raise SkipTest('Not implemented yet')
+        assert calculated_ranges == expected_ranges
 
-        calculated_ranges = self.config._get_search_ranges([self.rule], effect_clues)
+    def test_search_range_delayed_logs(self):
+        constraints1 = [
+            {
+                'clues_groups': [[1, 1], [0, 1]],
+                'name': 'time',
+                'params': {
+                    'min_delta': -10,
+                    'max_delta': 100
+                }
+            },
+        ]
+        rule = Rule([self.cause1, self.cause2], self.effect, constraints1, Rule.LINKAGE_AND)
+        calculated_ranges = self.calculate_range(rule)
 
         expected_ranges = {
             'database': {
                 'date': {
-                    'left_bound': datetime(2016, 4, 12, 23, 53, 3),
-                    'right_bound': datetime(2016, 4, 12, 23, 54, 33)
+                    'left_bound': self.one_hundred_second_earlier,
+                    'right_bound': self.ten_second_later
                 }
             },
+        }
+
+        raise SkipTest('Not implemented yet')
+        assert calculated_ranges == expected_ranges
+
+    def test_search_range_on_identical_constraint(self):
+        constraints1 = [{'clues_groups': [[1, 1], [0, 1]], 'name': 'identical', 'params': {}},]
+        rule = Rule([self.cause1, self.cause2], self.effect, constraints1, Rule.LINKAGE_AND)
+        calculated_ranges = self.calculate_range(rule)
+
+        expected_ranges = {
+            'database': {
+                'date': {
+                    'left_bound': self.effect_time,
+                    'right_bound': self.effect_time
+                }
+            },
+        }
+
+        raise SkipTest('Not implemented yet')
+        assert calculated_ranges == expected_ranges
+
+    def test_search_range_merge_range(self):
+        constraints1 = [
+            {
+                'clues_groups': [[1, 1], [0, 1]],
+                'name': 'time',
+                'params': {
+                    'max_delta': 75,
+                    'min_delta': 10
+                }
+            }, {
+                'clues_groups': [[2, 1], [0, 1]],
+                'name': 'time',
+                'params': {'max_delta': 10}
+            }, {
+                'clues_groups': [[3, 1], [0, 1]],
+                'name': 'time',
+                'params': {
+                    'max_delta': 100,
+                    'min_delta': 20
+                }
+            }, {
+                'clues_groups': [[3, 1], [0, 1]],
+                'name': 'time',
+                'params': {
+                    'max_delta': 100,
+                    'min_delta': 20
+                }
+            }
+        ]
+        rule = Rule(
+            [self.cause1, self.cause2, self.cause1, self.cause2], self.effect, constraints1,
+            Rule.LINKAGE_AND
```

Comment:
Ouch, the specified results seems to be correct when linkage=OR, but not AND. If some 'database' log occurs 80 secs before effect, it will be rejected by constraint no 1 (due to AND linkage between them), so it's useless to provide investigation in time range between 100 and 75 secs before effect.